FEXCore/Source/Interface/IR/Passes/RegisterAllocationPass.cpp

```diff
@@ -5,1239 +5,713 @@ tags: ir|opts
+    // At the start of each block, all registers are available.
+    for (auto& Class : Classes) {
+      Class.Available = (1u << Class.Count) - 1;
+      Class.Allocated = 0;
+    }
 
-      auto [InterferenceOrderedNode, _] = IR.at(InterferenceNode)();
-      auto InterferenceNodeOpBeginIter = IR.at(InterferenceLiveRange->Begin);
-      auto InterferenceNodeOpEndIter = IR.at(InterferenceLiveRange->End);
-
-      // If the nodes live range is entirely encompassed by the interference node's range
-      // then spilling that range will /potentially/ lower RA
-      // Will only lower register pressure if the interference node does NOT have a use inside of
-      // this live range's use
-      // | Our Node             |        Interference |
-      // | ========================================== |
-      // | 0                    |              Assign |
-      // | 1 - Assign           |            (No Use) |
-      // | 2                    |            (No Use) |
-      // | 3 - Last Use         |            (No Use) |
-      // | 4                    |                     |
-      // | 5                    |            Last Use |
-      // | Range - (1, 3]       |              (0, 5] |
-      if (CurrentLocation > InterferenceLiveRange->Begin && OpLiveRange->End < InterferenceLiveRange->End) {
-
-        // This will only save register pressure if the interference node
-        // does NOT have a use inside of this this node's live range
-        // Search only inside the source node's live range to see if there is a use
-        auto FirstUseLocation = FindFirstUse(IREmit, InterferenceOrderedNode, NodeOpBeginIter, NodeOpEndIter);
-        if (FirstUseLocation == IR::NodeIterator::Invalid()) {
-          // Looks like there isn't a usage of this interference node inside our node's live range
-          // This means it is safe to spill this node and it'll result in in lower RA
-          // Proper calculation of cost to spill would be to calculate the two distances from
-          // (Node->Begin - InterferencePrevUse) + (InterferenceNextUse - Node->End)
-          // This would ensure something will spill earlier if its previous use and next use are farther away
-          auto InterferenceNodeNextUse = FindFirstUse(IREmit, InterferenceOrderedNode, NodeOpBeginIter, InterferenceNodeOpEndIter);
-          auto InterferenceNodePrevUse = FindLastUseBefore(IREmit, InterferenceOrderedNode, InterferenceNodeOpBeginIter, NodeOpBeginIter);
-          LOGMAN_THROW_A_FMT(InterferenceNodeNextUse != IR::NodeIterator::Invalid(), "Couldn't find next usage of op");
-          // If there is no use of the interference op prior to our op then it only has initial definition
-          if (InterferenceNodePrevUse == IR::NodeIterator::Invalid()) {
-            InterferenceNodePrevUse = InterferenceNodeOpBeginIter;
-          }
+    // Next-use distance relative to the block end of each source, last first.
+    fextl::vector<uint32_t> SourcesNextUses;
```

Comment:
How many blocks does RA typically run over? Might make sense to move the declaration of this vector out of the loop and just `.clear()` it at the loop entry to minimize memory reallocations.